Feature/disabled hail

Elyssaen opened the pull request:
Each government can define a "hostile disabled hail" and "friendly
disabled hail", like "hostile hail" and "friendly hail". These are no
longer hard-coded. The game will still revert back to the same
default messages, i.e., Shakespearean insults for hostile, and "Our
ship has been disabled! ..." for friendly. Combines @EndrosG's
changes: see issues #2526 and #2531.

EndrosG committed:
friendlyDisabledHail

Elyssaen committed:
Added facility for a hostile disabled hail

Each government can define a "hostile disabled hail" like a "friendly
hail" or "hostile hail". If one if missing, the game will default the
"hostile disabled" phrases, i.e., Shakespearean insults.

Elyssaen committed:
Added the default friendly disabled hail to hails.txt

tehhowch reviewed (approved):
The only thing i might add would be a comment in HailPanel that indicates the "missing" conditions are handled by the default messages

Elyssaen commented:
@tehhowch, how would you word that?

tehhowch commented on source/HailPanel.cpp:
```diff
@@ -55,25 +55,18 @@ HailPanel::HailPanel(PlayerInfo &player, const shared_ptr<Ship> &ship)
 	else if(!hasLanguage)
 		message = "(An alien voice says something in a language you do not recognize.)";
-	else if(gov->IsEnemy())
+	else if(gov->IsEnemy() && !ship->IsDisabled())
```
Right here you could add
`// Update default hail responses based on the hailed ship's status conditions and government.`

Also, is there any repercussion from allowing hostile ships to fall down to the `else` condition? In the original, only friendly ships can make it that far

Elyssaen replied:
Oddly, it looks like it might be a problem but I haven't replicated it. I'm not in a coding state of mind right now so I'll have to leave it until I am.

EndrosG replied:
@Elyssaen most likely just wanted to fuse the two nested conditions `if(gov->IsEnemy())` and `if(!ship->IsDisabled())` into one after removing the `if(ship->IsDisabled())` branch inside the IsEnemy() branch.
The `IsDisabled()` branch gets removed because the new intention is that this function does not return any result in that case, and after removing it, only the `else` branch remains and thus it seemed logic to merge both nested if()s

That's also the reason why the next `else if()` had to be extended by `!gov->isEnemy()`, to prevent those ships being handled there.
In terms of a nice, readable, and understandable code, I agree with @tehhowch that a proper comment would be nice. Ideally, that comment just could replace the original `message = GameData:: ...` statement.
But then, we should also do the same with the `else message = "Our ship has been disabled! ..."` lines I removed for my friendlyDisabledHail part.

tehhowch replied:
Yes, it seems pretty unlikely to get a falldown, but if you are disabled/without fuel, and hail a hostile disabled ship, then you get this:
![image](https://cloud.githubusercontent.com/assets/20871346/26322285/e303e774-3eef-11e7-9c90-b4e071245efb.png)

tehhowch replied:
You can see he's disabled in the target reticle, and that I also am in my flagship reticle. While options to request assistance are greyed out, this situation's falldown is probably not acceptable to MZ

tehhowch committed:
Address corner case when hailing hostiles.

If the selected hostile ship is disabled, and the player flagship needs fuel/repairs, the "assistance" hail panel would be shown instead of a hostile hail.
Also adds a comment regarding the default hail messages.

EndrosG commented on source/HailPanel.cpp:
```diff
@@ -55,25 +55,22 @@ HailPanel::HailPanel(PlayerInfo &player, const shared_ptr<Ship> &ship)
 		}
 	}
-	else if(ship->IsDisabled())
+	else if(!gov->IsEnemy() && ship->IsDisabled())
```
the `!gov->IsEnemy() && ` is obsolete again, it was only neccessary in the temporary state with the `if(gov->IsEnemy() && !ship->IsDisabled())` clause.

tehhowch replied:
:+1:

I didn't catch that, oops

Elyssaen replied:
Sure about this? If you remove `!gov->isEnemy() &&`, then if:

1. You have no JumpsRemaining, or you're disabled, and
2. They're an enemy, and
3. They're disabled

surely they'll reply "Sorry, we can't help you, because our ship is disabled". But they should reply with their hostile disabled hail.

_Edit:_ ah, no. That would've been true based on how I had the above block originally, but that changed, didn't it...
_Edit 2_: I _really_ don't like what we're doing to this whole block, actually. I think it's more a function of how I don't much like how the block worked in the first place, and I feel like we're fiddling with it by halves. `else if (blah) { if(bleh) } else ...` really frustrates me stylistically.

tehhowch replied:
Yes it did. All hostile ships are caught in line 59

Elyssaen replied:
Sorry, got an edit in before that! At one point I'd rewritten the above block to be

`else if(gov->IsEnemy() && !ship->IsDisabled())`

and I didn't catch that we'd changed it back.

Elyssaen commented:
I think my key worry about what we're doing here (and why I more wanted to _just_ do hostile disabled, minimally) is that we're fiddling about with the HailPanel block in a situation where, if it was our own project, we'd just rewrite the whole thing. 

It'd be a good opportunity to rewrite it all to be clearer, more localisation-friendly, and to put *all* the text responses onto ship->GetHail() or government->GetHail(). (And maybe enhance ship->GetHail() to not simply give one message regardless of circumstance.) But while that might be one owner's commit, it's not one contributor's PR. Just seems like a tricky situation.

Elyssaen committed:
Removed an unnecessary check

tehhowch commented:
⚠️ suggestions based on incomplete hail structure understanding follow ⚠️ 

You could branch from 016d709 (e.g. assume it merges 🙊 ) and perform the refactor. My guess is that Ship::GetHail() was designed for general comms use, and HailPanel is for special responses that are exclusively relevant to the player.

I could see Ship::GetHail() taking an argument `isHailPanel`, branching on "Am I disabled?" and `isHailPanel`, and requesting `Government::GetHail(isHailPanel, amDisabled).
Gov::GetHail then checks the ship's hostility to the player (isHail) or the system government (notHail), and selects the appropriate friendly/hostile disabled/notdisabled hail/retort.
